src/V3Delayed.cpp

```diff
@@ -76,140 +114,174 @@
-        AstNodeExpr* rhsp = nullptr;  // The captured value being assigned by the AstAssignDly
-        AstActive* activep = nullptr;  // The active block the AstAssignDly is under
-        AstNode* insertionPointp = nullptr;  // The insertion point after the AstAssignDly
-        std::string suffix;  // Suffix to add to additional variables created
-        bool consecutive = false;  // Record corresponds to an AstAssignDly immediately adjacent
+    struct NBA final {
+        AstAssignDly* nodep = nullptr;  // The NBA this record refers to
+        AstVarScope* vscp = nullptr;  // The target variable the NBA is updating
     };
 
     // NODE STATE
-    //  AstVarScope::user1p()   -> aux
-    //  AstVar::user1()         -> bool.  Set true if already made warning
-    //  AstVarRef::user1()      -> bool.  Set true if is a non-blocking reference
-    //  AstAlwaysPost::user1()  -> AstIf*.  Last IF (__VdlySet__) created under this AlwaysPost
-    //
-    // Cleared each scope/active:
-    //  AstVarScope::user2()      -> AstVarScope*.  Commit queue for this variable
-    //  AstVarRef::user2()        -> bool.  Set true if already processed
-    //  AstAssignDly::user2()     -> AstVarScope*.  __VdlySet__ created for this assign
-    //  AstAlwaysPost::user2()    -> AstVarScope*.  __VdlySet__ last referenced in IF
-    //  AstNodeProcedure::user2() -> AstAcive*.  Active block used by this suspendable process
-    const VNUser1InUse m_inuser1;
-    const VNUser2InUse m_inuser2;
-
-    struct AuxAstVarScope final {
-        // Points to temp (shadow) variable created.
-        AstVarScope* delayVscp = nullptr;
-        // Points to AstActive block of the shadow variable 'delayVscp/post block 'postp'
-        AstActive* activep = nullptr;
-        // Post block for this variable
-        AstAlwaysPost* postp = nullptr;
-        // First reference encountered to the VarScope
-        const AstNodeVarRef* firstRefp = nullptr;
-        // If the implementation cannot be deferred, this is the location why
-        FileLine* nonDeferredFlp = nullptr;
-        // Variable requires a commit queue due to dynamic NBA
-        bool needsCommitQueue = false;
-        // Array element is partially updated by at least some NBA
-        bool hasPartialUpdate = false;
-    };
-    AstUser1Allocator<AstVarScope, AuxAstVarScope> m_vscpAux;
+    //  AstVar::user1()         -> bool.  Set true if already issued MULTIDRIVEN warning
+    //  AstVarRef::user1()      -> bool.  Set true if target of NBA
+    //  AstNodeModule::user1p() -> std::unorded_map<std::string, AstVar*> temp map via m_varMap
+    //  AstVarScope::user1p()   -> VarScopeInfo via m_vscpInfo
+    //  AstVarScope::user2p()   -> AstVarRef*: First write reference to the Variable
+    const VNUser1InUse m_user1InUse{};
+    const VNUser2InUse m_user2InUse{};
+    AstUser1Allocator<AstNodeModule, std::unordered_map<std::string, AstVar*>> m_varMap;
```

Comment:
```suggestion
    AstUser1Allocator<AstNodeModule, std::map<std::string, AstVar*>> m_varMap;
```

Having the large unordered structure per-module will otherwise cause perf problems if many small modules.

Reply:
Can you explain what you think the issue is here? I don't see one, so I'm curious what I am missing.

Reply:
Older libc++'s allocate and clear dozens of KB for unordered_set/unordered_map.  So they are much slower when small and frequently constructed, so best used only for e.g. one-per-netlist not one-per-module. I didn't search for the point when this was improved, but it is ok in libc++11, so we can probably ignore this once we require C++20.

Reply:
From GCC 5.1 (~April 2015):
https://godbolt.org/z/5zo1nYYWv
From Clang 3.7 (~September 2015):
https://godbolt.org/z/cncsveYGs

So has not been an issue for ~9 years. Would be remarkable if the code base built with those compilers, but if you still care I will change it :)

Reply:
It was still an issue in 2021. This change gave a very big boost. 8b00939f0cd0c9e832ceaae55a956e343bfb2320

Reply:
https://github.com/verilator/verilator/blob/8b00939f0cd0c9e832ceaae55a956e343bfb2320/src/V3Scoreboard.h#L368-L369

>  Since the set size is generally small,

FWIW I think that's the real reason why that patch helped you.

unordered_set.clear does not reduce the bucket count, so this is the problem use case:

```
std::unorderd_set set;
while (...) {
  addSomeElements(set);
  set.clear();
}
```

Now, if most of the time `addSomeElements` adds a small number of elements, the clear is proportional. However, if on one iteration, you add a lot more elements, the clear on all subsequent invocation will be expensive (clear is O(set.bucket_count())), as the reserved size of the set is never reduced. This is pretty easy to avoid with:

```
std::unorderd_set set;
while (...) {
  addSomeElements(set);
  counst double lf = set.load_factor();
  set.clear();
  // Sssuming set.max_load_factor() == 1.0
  if (lf < 0.25) set.reserve(set.bucket_count() / 2);
}
```

Note this only happens when you repeatedly call `clear` on the same instance, which we don't do here.

Just FYI.